common/src/main/java/com/wynntils/core/text/StyledTextPart.java

```diff
@@ -0,0 +1,139 @@
+import net.minecraft.ChatFormatting;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.network.chat.Style;
+
+public final class StyledTextPart {
+    private final String text;
+    private PartStyle style;
+
+    private final StyledText parent;
+
+    StyledTextPart(String text, Style style, StyledText parent, Style parentStyle) {
+        this.parent = parent;
+        this.text = text;
+
+        // Must be done last
+        this.style = PartStyle.fromStyle(style, this, parentStyle);
+    }
+
+    StyledTextPart(StyledTextPart part, StyledText parent) {
+        this.text = part.text;
+        this.style = part.style;
+        this.parent = parent;
+    }
+
+    // This factory is used to create a StyledTextPart from a string that has formatting codes
+    // It is separate from the constructor because this only needs to be applied in cases there the text could have
+    // formatting codes
+    static StyledTextPart fromStyledString(String text, Style style, StyledText parent, Style parentStyle) {
+        // When we have a style, but the text has formatting codes,
+        // we need to apply the formatting codes to the style
+        // This means that the actual style applies first; then the formatting codes
+        StringBuilder textBuilder = new StringBuilder(text.length());
+        Style textStyle = style;
+        boolean formattingNext = false;
+        for (char current : text.toCharArray()) {
+            if (formattingNext) {
+                formattingNext = false;
+                ChatFormatting formatting = ChatFormatting.getByCode(current);
+
+                if (formatting == null) {
+                    textBuilder.append(ChatFormatting.PREFIX_CODE);
+                    textBuilder.append(current);
+                    continue;
+                }
+
+                // Color formatting resets the style
+                if (formatting.isColor()) {
+                    textStyle = Style.EMPTY.withColor(formatting);
+                } else {
+                    textStyle = textStyle.applyFormat(formatting);
+                }
+
+                continue;
+            }
+
+            if (current == '§') {
+                formattingNext = true;
+                continue;
+            }
+
+            textBuilder.append(current);
+        }
+
+        // We might have lost an event, so we need to add it back
+        textStyle = textStyle.withClickEvent(style.getClickEvent()).withHoverEvent(style.getHoverEvent());
+
+        return new StyledTextPart(textBuilder.toString(), textStyle, parent, parentStyle);
+    }
+
+    public String getString(PartStyle previousStyle, PartStyle.StyleType type) {
+        return style.asString(previousStyle, type) + text;
+    }
+
+    public StyledText getParent() {
+        return parent;
+    }
+
+    public PartStyle getPartStyle() {
+        return style;
+    }
+
+    public void setPartStyle(PartStyle style) {
```

Comment:
I do not like this. Why do you think our parts should be mutable? I'd rather see a "wither". 

I can possibly live with our StyledText being mutable, so we can swap out parts etc (like a list of immutable objects), but I don't think the parts or the style should be mutable.

And I'm frankly not sure that StyledText should be mutable as well. There is a case to be made for saying that if we want to modify a StyledText, we should replace it with a new, modified StyledText instead. I'm not sure about that one yet...

Reply:
StyledText is immutable. This is the only part where we have some kind of mutability. I can't see any way of creating an API for "I want to change this parts style without actually leaking this part". Well, we could have `findMatchingPartAndReplaceStyleWith` but that is kind of weird..

Reply:
But why should you change an existing part? You can create a new part based on an existing part + the new style that you want.

And then you can create a new StyledText based on the old one, but where the old part of interest was, is now the new part you have created.

(This is actually the one place where I might think mutability can be allowed -- that you can, like a List, be allowed to change what parts a StyledText consists of). 

That way you could do like:
```
int index = 0;

do {
index = styledText.findMatch(myRegexp, index);
if (index == -1) return;
StyledTextPart matchingPart = styledText.getPart(index);
StyledTextPart modifiedPart = fixup(matchingPart);
styledText.setPart(index, modifiedPart);
} while (true);
```

Reply:
That is, basically let StyledTextPart implement Collection (not sure if we should *actually* do this, but at least those operations that makes sense).

Reply:
I think I did this now. I used `replacePart` instead of implementing custom methods. Please check if you are happy.